### PR TITLE
Rework use-images-navigation hook tests

### PR DESCRIPTION
**Pull request by Jordanlelay**
## Work performed

- Rework use-images-navigation hook tests

### CODE REVIEW

File: typescript/web/src/hooks/__tests__/use-images-navigation.tsx [hunk 1]
```diff
@@ -1,90 +1,58 @@
...
-const Wrapper = ({ children }: React.PropsWithChildren<{}>) => (
-  <ApolloProvider client={client}>{children}</ApolloProvider>
+const ImageNotInDatasetWrapper = getApolloMockWrapper(
+  getApolloMockLink(CURRENT_NOT_IN_IMAGES_MOCKS)
 );
 
 test("The currentImageIndex is undefined while loading", async () => {
   const { result } = renderHook(() => useImagesNavigation(), {
-    wrapper: Wrapper,
+    wrapper: NoImagesWrapper,
```
Comment (bbenoist) [inline typescript/web/src/hooks/__tests__/use-images-navigation.tsx:45]: Instead of declaring custom wrappers for each variant, you could factorize a little bit more and have a ` renderTest` function looking like this one:

```tsx
const renderTest = (mocks: ApolloMockResponses) => {
  const link = getApolloMockLink(mocks);
  const wrapper = getApolloMockWrapper(link);
  return renderHook(() => useImagesNavigation(), { wrapper });
};
```

Then, in each test, just call something like:

```ts
const { result, waitForValueToChange } = renderTest(CURRENT_NOT_IN_IMAGES_MOCKS);
```

File: typescript/web/src/hooks/use-image-navigation.fixtures.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  ApolloMockResponse,
+  ApolloMockResponses,
+} from "../utils/tests/apollo-mock";
+import { BASIC_DATASET_DATA } from "../utils/tests/data.fixtures";
+import { CURRENT_IMAGE } from "./test-utils";
+
+export const IMAGE1 = {
+  id: "929f5445-812c-4679-86bd-0f0748679f38",
+  url: "",
+  thumbnail200Url: "",
```
Comment (bbenoist) [inline typescript/web/src/hooks/use-image-navigation.fixtures.ts:17]: Can we at-least use something which would work?
For example something like `https://labelflow.ai/static/icon-512x512.png`.

File: typescript/web/src/hooks/use-image-navigation.fixtures.ts [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  getImagesMockGenerator(CURRENT_IMAGE_IS_LAST),
+];
+
+export const NO_IMAGES_MOCKS: ApolloMockResponses = [
+  getImagesMockGenerator([]),
+];
+
+export const CURRENT_NOT_IN_IMAGES_MOCKS: ApolloMockResponses = [
+  getImagesMockGenerator(CURRENT_IMAGE_NOT_PRESENT),
+];
```
Comment (bbenoist) [inline typescript/web/src/hooks/use-image-navigation.fixtures.ts:81]: Can you add an intermediary `createImagesMocks` function which would return directly the array?

File: typescript/web/src/hooks/use-image-navigation.fixtures.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+const CURRENT_IMAGE_IS_SECOND = [IMAGE1, CURRENT_IMAGE, IMAGE2];
+const CURRENT_IMAGE_IS_LAST = [IMAGE1, IMAGE2, CURRENT_IMAGE];
+const CURRENT_IMAGE_NOT_PRESENT = [IMAGE1, IMAGE2, IMAGE3];
+
+const getImagesMockGenerator = (
+  images: GetAllImagesOfADatasetQuery_dataset_images[]
+): ApolloMockResponse<
+  GetAllImagesOfADatasetQuery,
+  GetAllImagesOfADatasetQueryVariables
+> => {
```
Comment (bbenoist) [inline typescript/web/src/hooks/use-image-navigation.fixtures.ts:42]: The braces are useless here

File: typescript/web/src/hooks/use-image-navigation.fixtures.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  url: "",
+  thumbnail200Url: "",
+};
+
+const CURRENT_IMAGE_IS_FIRST = [CURRENT_IMAGE, IMAGE1, IMAGE2];
+const CURRENT_IMAGE_IS_SECOND = [IMAGE1, CURRENT_IMAGE, IMAGE2];
+const CURRENT_IMAGE_IS_LAST = [IMAGE1, IMAGE2, CURRENT_IMAGE];
+const CURRENT_IMAGE_NOT_PRESENT = [IMAGE1, IMAGE2, IMAGE3];
+
+const getImagesMockGenerator = (
```
Comment (bbenoist) [inline typescript/web/src/hooks/use-image-navigation.fixtures.ts:37]:
```suggestion
const createImagesMock = (
```

File: typescript/web/src/hooks/use-image-navigation.fixtures.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    },
+    result: {
+      data: {
+        dataset: {
+          id: BASIC_DATASET_DATA.id,
+          images,
+        },
+      },
+    },
+    nMatches: Number.POSITIVE_INFINITY,
```
Comment (bbenoist) [inline typescript/web/src/hooks/use-image-navigation.fixtures.ts:59]: The mocked query is not a function. Why is that needed exactly?

**Review by bbenoist — CHANGES_REQUESTED** [review]

File: typescript/web/src/hooks/use-image-navigation.fixtures.ts (hunk 6 above)
Comment (Jordanlelay) [reply typescript/web/src/hooks/use-image-navigation.fixtures.ts:59]: Are you talking about `nMatches` property ? If yes, I don't understand the `query is not a function` part

It's needed because the mutation is used in parallel. If you don't specify it, you will receive a warning saying that the mutation is not mocked after the first call

File: typescript/web/src/hooks/test-utils.ts [hunk 7]
```diff
@@ -0,0 +1,12 @@
+/**
+ * This constant is not exported directly from
+ * `typescript/web/src/hooks/use-image-navigation.fixtures.ts`
+ * otherwise, the `mockNextRouter` function does not work, making all
+ * tests in `typescript/web/src/hooks/__tests__/use-images-navigation.tsx`
+ * fail
```
[the remark this answers is not in the document]
Comment (Jordanlelay) [reply typescript/web/src/hooks/test-utils.ts:6]: It happens in many tests, we have to mock the router before some imports. In my case, the router has to be mocked before the mocked query in the `.fixtures.ts` file.
And because I wanted to use an imported variable in my mocked router, I had to extract it from my `.fixtures.ts` file

File: typescript/web/src/hooks/use-image-navigation.fixtures.ts (hunk 3 above)
Comment (Jordanlelay) [reply typescript/web/src/hooks/use-image-navigation.fixtures.ts:81]: Which array ?

Comment (bbenoist) [reply typescript/web/src/hooks/use-image-navigation.fixtures.ts]: The `[` and `]` surrounding every value declared in the block I've mentioned.

File: typescript/web/src/hooks/use-image-navigation.fixtures.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+export const IMAGE_3_DATA = {
+  id: "8067a965-d017-4c4b-a82d-98cc803d1020",
+  url: "https://labelflow.ai/static/icon-512x512.png",
+  thumbnail200Url: "https://labelflow.ai/static/icon-512x512.png",
+};
+
+const CURRENT_IMAGE_IS_FIRST = [CURRENT_IMAGE, IMAGE_1_DATA, IMAGE_2_DATA];
+const CURRENT_IMAGE_IS_SECOND = [IMAGE_1_DATA, CURRENT_IMAGE, IMAGE_2_DATA];
+const CURRENT_IMAGE_IS_LAST = [IMAGE_1_DATA, IMAGE_2_DATA, CURRENT_IMAGE];
+const CURRENT_IMAGE_NOT_PRESENT = [IMAGE_1_DATA, IMAGE_2_DATA, IMAGE_3_DATA];
```
Comment (bbenoist) [inline typescript/web/src/hooks/use-image-navigation.fixtures.ts:35]:
```suggestion

const EXAMPLE_URL = "https://labelflow.ai/static/icon-512x512.png";

const IMAGES_IDS = [
  "dd5f0305-3d69-457d-9a77-4a46cec0e10b",
  "929f5445-812c-4679-86bd-0f0748679f38",
  "51e00f79-d178-4ea1-8bb5-b0fa647f9d19",
  "8067a965-d017-4c4b-a82d-98cc803d1020",
];

export const [CURRENT_IMAGE_DATA, IMAGE_1_DATA, IMAGE_2_DATA, IMAGE_3_DATA] =
  IMAGES_IDS.map((id) => ({
    id,
    url: EXAMPLE_URL,
    thumbnail200Url: EXAMPLE_URL,
  }));

const CURRENT_IMAGE_IS_FIRST = [CURRENT_IMAGE_DATA, IMAGE_1_DATA, IMAGE_2_DATA];
const CURRENT_IMAGE_IS_SECOND = [IMAGE_1_DATA, CURRENT_IMAGE_DATA, IMAGE_2_DATA];
const CURRENT_IMAGE_IS_LAST = [IMAGE_1_DATA, IMAGE_2_DATA, CURRENT_IMAGE_DATA];
const CURRENT_IMAGE_NOT_PRESENT = [IMAGE_1_DATA, IMAGE_2_DATA, IMAGE_3_DATA];
```

File: typescript/web/src/hooks/__tests__/use-images-navigation.tsx [hunk 9]
```diff
@@ -1,90 +1,49 @@
...
+import { CURRENT_IMAGE } from "../test-utils";
+
+mockNextRouter({
+  isReady: true,
+  query: {
+    imageId: CURRENT_IMAGE.id,
+    datasetSlug: BASIC_DATASET_DATA.slug,
+    workspaceSlug: BASIC_DATASET_DATA.workspace.slug,
+  },
+});
```
Comment (bbenoist) [inline typescript/web/src/hooks/__tests__/use-images-navigation.tsx:14]:
```suggestion
```

File: typescript/web/src/hooks/__tests__/use-images-navigation.tsx [hunk 10]
```diff
@@ -1,90 +1,49 @@
...
-  })),
-}));
+  CURRENT_NOT_IN_IMAGES_MOCKS,
+  IMAGE_1_DATA,
+  IMAGE_2_DATA,
+  IMAGE_IS_FIRST_MOCKS,
+  IMAGE_IS_LAST_MOCKS,
+  NO_IMAGES_MOCKS,
+  THREE_IMAGES_MOCKS,
+} from "../use-image-navigation.fixtures";
```
Comment (bbenoist) [inline typescript/web/src/hooks/__tests__/use-images-navigation.tsx:23]:
```suggestion
  CURRENT_IMAGE_DATA,
  CURRENT_NOT_IN_IMAGES_MOCKS,
  IMAGE_1_DATA,
  IMAGE_2_DATA,
  IMAGE_IS_FIRST_MOCKS,
  IMAGE_IS_LAST_MOCKS,
  NO_IMAGES_MOCKS,
  THREE_IMAGES_MOCKS,
} from "../use-image-navigation.fixtures";

mockNextRouter({
  isReady: true,
  query: {
    imageId: CURRENT_IMAGE_DATA.id,
    datasetSlug: BASIC_DATASET_DATA.slug,
    workspaceSlug: BASIC_DATASET_DATA.workspace.slug,
  },
});
```

**Review by bbenoist — CHANGES_REQUESTED** [review]

File: typescript/web/src/hooks/use-image-navigation.fixtures.ts (hunk 8 above)
Comment (Jordanlelay) [reply typescript/web/src/hooks/use-image-navigation.fixtures.ts]: Looks a bit much for a test like that

Comment (bbenoist) [reply typescript/web/src/hooks/use-image-navigation.fixtures.ts]: It's common factorization... At-least it takes less space (_don't forget CURRENT_IMAGE_ 😉) and don't duplicate the URLs everywhere...

File: typescript/web/src/hooks/use-image-navigation.fixtures.ts (hunk 6 above)
Comment (bbenoist) [reply typescript/web/src/hooks/use-image-navigation.fixtures.ts]: Yes, I'm talking about `nMatches`. Restricting a query mock to a certain number of calls adds a layer of trust. It lets you be sure that the hook don't overcall the query or that the skip property is honored...

> I don't understand the query is not a function part

Yes, I was unclear. I meant "the >mocked< query >response< is not a function". Because, the most common case where you'd want to set nMatches to `Number.POSITIVE_INFINITY` is when you've implemented the mocked query return value as a function instead of static value.

To follow-up my question, why is `GET_ALL_IMAGES_OF_A_DATASET_QUERY` be called in parallel exactly?

Comment (Jordanlelay) [reply typescript/web/src/hooks/use-image-navigation.fixtures.ts]: Actually what I said was wrong, this mock does not need this nMatches property. I was confused because of my 2 other PRs I guess 😄

**Review by bbenoist — APPROVED** [review]
Good job! And we end-up with less code than before!
image
